### PR TITLE
Add command to filter residents by specific fields

### PR DESCRIPTION
**Pull request by nareus**

### CODE REVIEW

File: src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            + PREFIX_EMAIL + "johndoe@example.com";
+
+    public static final String MESSAGE_FILTER_PERSON_SUCCESS = "filtered Person: %1$s";
+    public static final String MESSAGE_NOT_FILTERED = "At least one field to filter must be provided.";
+    public static final String MESSAGE_DUPLICATE_PERSON = "This person already exists in the address book.";
+
+    private final FilterPersonDescriptor filterPersonDescriptor;
+
+    /**
+     * @param filterPersonDescriptor details to filter the person with
```
Comment (alvinjiang1) [inline src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java:52]: I would reuse the word "description" here, as well as standardise the target as "resident", so "description to filter the resident with".

File: src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        private Address address;
+        private Set<Tag> tags;
+
+        public FilterPersonDescriptor() {}
+
+        /**
+         * Copy constructor.
+         * A defensive copy of {@code tags} is used internally.
+         */
+        public FilterPersonDescriptor(FilterCommand.FilterPersonDescriptor toCopy) {
```
Comment (alvinjiang1) [inline src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java:87]: I don't think we need to say that it's a `Copy constructor`. And do also consider using imperative mood, so something like `Constructs a ...`

File: src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            setPhone(toCopy.phone);
+            setEmail(toCopy.email);
+            setAddress(toCopy.address);
+            setTags(toCopy.tags);
+        }
+
+        /**
+         * Returns true if at least one field is filtered.
+         */
+        public boolean isAnyFieldfiltered() {
```
Comment (alvinjiang1) [inline src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java:98]: Missing `@return` in method header.

**Review by nealetham — APPROVED** [review]
Reviewed the code quality. Good usage of SLAP 👍

File: src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+        public Address getAddress() {
+            return address;
+        }
+
+        /**
+         * Sets {@code tags} to this object's {@code tags}.
+         * A defensive copy of {@code tags} is used internally.
+         */
+        public void setTags(Set<Tag> tags) {
```
Comment (alvinjiang1) [inline src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java:138]: Same thing as above: use imperative mood + `@param`

File: src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            + "[" + PREFIX_EMAIL + "EMAIL] "
+            + "[" + PREFIX_ADDRESS + "ADDRESS] "
+            + "[" + PREFIX_TAG + "TAG]...\n"
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_PHONE + "91234567 "
+            + PREFIX_EMAIL + "johndoe@example.com";
+
+    public static final String MESSAGE_FILTER_PERSON_SUCCESS = "filtered Person: %1$s";
+    public static final String MESSAGE_NOT_FILTERED = "At least one field to filter must be provided.";
+    public static final String MESSAGE_DUPLICATE_PERSON = "This person already exists in the address book.";
```
Comment (jq1836) [inline src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java:47]: Is this really a problem that could be encountered by filtering? Seems more like a problem that would happen when adding a person to the list. Could we possibly remove this?

File: src/main/java/seedu/rc4hdb/model/person/AttributesMatchKeywordsPredicate.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import static java.util.Objects.isNull;
+
+import java.util.function.Predicate;
+
+import seedu.rc4hdb.logic.commands.modelcommands.FilterCommand;
+
+/**
+ * Tests that a {@code Person}'s {@code Name} matches any of the keywords given.
+ */
+public class AttributesMatchKeywordsPredicate implements Predicate<Person> {
```
Comment (jq1836) [inline src/main/java/seedu/rc4hdb/model/person/AttributesMatchKeywordsPredicate.java:12]: Parts of this class could be further abstracted, let's split this up into smaller predicates, which will make up this larger predicate.

File: src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        public void setTags(Set<Tag> tags) {
+            this.tags = (tags != null) ? new HashSet<>(tags) : null;
+        }
+
+        /**
+         * Returns an unmodifiable tag set, which throws {@code UnsupportedOperationException}
+         * if modification is attempted.
+         * Returns {@code Optional#empty()} if {@code tags} is null.
+         */
+        public Set<Tag> getTags() {
```
Comment (alvinjiang1) [inline src/main/java/seedu/rc4hdb/logic/commands/modelcommands/FilterCommand.java:147]: Same thing as above: use `@return`

File: src/main/java/seedu/rc4hdb/model/person/AttributesMatchKeywordsPredicate.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            predicate = predicate && person.getPhone().equals(descriptor.getPhone());
+        }
+        if (!isNull(descriptor.getEmail())) {
+            predicate = predicate && person.getEmail().equals(descriptor.getEmail());
+        }
+        if (!isNull(descriptor.getTags())) {
+            predicate = predicate && person.getTags().equals(descriptor.getTags());
+        }
+        return predicate;
+
```
Comment (jq1836) [inline src/main/java/seedu/rc4hdb/model/person/AttributesMatchKeywordsPredicate.java:38]: This is the section which could benefit greatly from the abstraction.

**Review by jq1836 — APPROVED** [review]
LGTM, apart from some classes that could be further broken down into other abstractions (which we can do later on).

**Review by alvinjiang1 — APPROVED** [review]
Javadocs generally look okay, just some minor nits. As Jordan previously pointed out, there could be some issues with method headers due to previous AB3 docucmentation. I would suggest editing based on my comments for now, and I would take a look afterwards to make sure everything is uniform afterwards. 

I'll open an issue for the remaining Javadocs improvements + updating of UG/DG.
Otherwise LGTM! Great work Naren :))

**Pull request merged**